### PR TITLE
Keybord nevigation Issue#10

### PR DESCRIPTION
**Pull request by JBaba**
Two cases are addressed in this patch which is raised in Issue#10,
1. Case No 1 to provide traversing of search link based on Up and Down key press event I have added "tabIndex" attribute in html page and index.js where search links are being rendered. Two functions in Index.js has been added which are "traverseElementOnKeyDown(event)" and "findElementByAttributeValue(e,v)".
2. Case No 2 by pressing "Esc"  key send focus back to html page also has been addressed.

### CODE REVIEW

File: templates/index.html [hunk 1]
```diff
@@ -68,11 +68,11 @@
 
     {{if getConfig.IsDemo}}
       <div id="demowarning">
-        Welcome! This is a <b>demo</b> of the <a href="https://about.commonsearch.org">Common Search</a> interface.<br/><br/>
+        Welcome! This is a <b>demo</b> of the <a href="https://about.commonsearch.org" tabindex="6">Common Search</a> interface.<br/><br/>
```
Comment (sylvinus) [inline templates/index.html]: I think we should have tabindex=-1 on all the elements of the demowarning?

File: static/js/index.js [hunk 2]
```diff
@@ -214,13 +256,16 @@
...
+	var tabIndexCount = 9;
+	
     var html = "";
     for (var i = 0; i < (result["h"] || []).length; i++) {
       var hit = result["h"][i];
       html += "<div class='r'>" +
-                "<h3><a href='"+hit["u"]+"'>"+htmlSafe(hit["t"])+"</a></h3>" +
-                "<div class='u'><a href='"+hit["u"]+"'>" + simplifyURL(hit["u"]) + "</a></div>" +
+                "<h3><a href='"+hit["u"]+"' tabindex='"+(tabIndexCount+=1)+"'>"+htmlSafe(hit["t"])+"</a></h3>" +
+                "<div class='u'><a href='"+hit["u"]+"' tabindex='"+(tabIndexCount+=1)+"'>" + simplifyURL(hit["u"]) + "</a></div>" +
```
Comment (sylvinus) [inline static/js/index.js]: This is the same link as in the H3 so we should bypass it?

File: static/js/index.js [hunk 3]
```diff
@@ -87,6 +100,35 @@
...
+	
+	// Find active Element tabIndex
+	var tabIndexElement = document.activeElement.tabIndex;
+	
+	if(tabIndexElement < 8){
+		return;
+	}
+	// Based on Up and Down Key press move focus from next or previous search links
+	if(event.keyCode == 38)	{ // up key press
+		findElementByAttributeValue('tabIndex',--tabIndexElement).focus();
```
Comment (sylvinus) [inline static/js/index.js]: I think we can use document.querySelector("[tabindex=N]") directly; We don't need JS support for IE<8

**Comment by sylvinus** [comment]
Thanks @JBaba for the PR! Looks very good. I've added a couple comments!

File: templates/index.html (hunk 1 above)
Comment (JBaba) [inline templates/index.html]: I will make changes for this section.

File: static/js/index.js (hunk 2 above)
Comment (JBaba) [inline static/js/index.js]: If we don't have tabIndex then how find next element. Or I don't get what you are trying to say.

Comment (sylvinus) [inline static/js/index.js]: Sorry if that wasn't clear! tabindex on the h3.a is great. But I don't think we need it on the second one (`<div class='u'>`). Each "tab" keypress should take you to the next result!

Comment (JBaba) [inline static/js/index.js]: I see i will make that change.

**Comment by sylvinus** [comment]
This works great!

A couple things I think we could improve:
- Add a better style for active links. Maybe just `text-decoration:underline`?
- When using the arrow keys, the page scrolls a bit each time. This doesn't happen with the "tab" key. I think we should avoid this behaviour (a `return false` or `event.preventDefault()` may do the trick?)
- When you arrive at the end of the list, the next "tab" keypress goes back on top, to the first `<div class="u">` element. Maybe instead of giving a specific tabindex to each h3.a, we could just explicitly give tabindex=-1 to the `<div class="u">` elements?

**Comment by JBaba** [comment]
I was about to address some of this that's great that you just mentioned. I
didn't know how could I address this changes while I was in middle of
development.
I will work on this addressed changes.

**Comment by JBaba** [comment]
@sylvinus Can you help me understand this "next" and "previous" behaviour because after clicking on either of this "renderHit" function is not being invoked and which results into tabIndex not being assigned to search links. I don't know what's happening here ?

**Comment by sylvinus** [comment]
Currently the "next" links are regular links that cause a page reload. So the page you get has been rendered from the Go template, not from JavaScript's `renderHit()`

**Comment by JBaba** [comment]
So only first page search result will use JavaScript's `renderHit()`. What's solution for rest of the pages and how can I help ? 
Even if you come back from second page to first by clicking on `previous` then also we don't have `tabIndex`

**Comment by sylvinus** [comment]
@JBaba it's the other way around: all results _except_ the first page load use renderHit().

You have already patched index.html to add some tabindexes, you just need to make the same changes there are you make in renderHit!

**Comment by JBaba** [comment]
I have tried to use `$index` but it always gives me `0,1,2,3....` series to start at my desired value what can I do.
Let me know your views I have never used GoLang before.

**Comment by sylvinus** [comment]
Can't you do `{{ add $index offset}}` ?

**Comment by JBaba** [comment]
@sylvinus  For this we have to define add function. I was reading similar article on this.

**Comment by sylvinus** [comment]
If you need to do that you can add it here:
https://github.com/commonsearch/cosr-front/blob/master/server/templates.go#L83

**Comment by JBaba** [comment]
Thanks that worked.

**Comment by JBaba** [comment]
I have addressed new changes.
Need help with scrolling page to top when we reach at end of screen but `preventDefault()` interrupts that.
I have tried using this code,

```
var h = window.innerHeight;
var rect = currentElement.getBoundingClientRect();
var offset = h - rect.top;
if(offset < 100 ){
    window.scrollTo(0, h/2);
} 
```

But this works for sometimes and sometimes not. Couldn't figure out why ?

**Comment by sylvinus** [comment]
Scrolling to `h/2` seems suspicious to me.. Shouldn't it be `rect.top`?

**Comment by JBaba** [comment]
This what i have found after referring few posts. I am finding offset (bottom) by subtracting document inner width from top of element. I was also not sure and after tested for while felt that this is buggy so asked for helped. What do you suggest how should we approach this ?

**Comment by sylvinus** [comment]
It seems like the general idea should work, but the formula doesn't look right. Try `window.scrollTo(0, rect.top);`?

**Comment by JBaba** [comment]
@sylvinus  I have simulated problem here you can see what's happening,
[https://jsfiddle.net/naimish101/bat8bf8h/4/](https://jsfiddle.net/naimish101/bat8bf8h/4/)

**Comment by sylvinus** [comment]
@JBaba Seems to be fixed if you use `offset < 0` instead ?

**Comment by JBaba** [comment]
@sylvinus you must be testing on chrome. It does not work vary well with other browsers.

**Comment by sylvinus** [comment]
It seems to work for me in Firefox as well. 

I'm working on interface tests for #1 so I'm going to merge it now and with Saucelabs we will make sure it behaves the same way in all browsers!

Thanks for the great work!

**Pull request merged**